rfc/rfc-0000-multiple-paths.md

```diff
@@ -0,0 +1,241 @@
+metadata:
+  name: supply-chain
+spec:
+  resources:
+    - name: source-provider
+      templateRef:
+        kind: ClusterSourceTemplate
+        options:
+          - name: source-from-git-repo
+            selector:
+              matchFields:
+                { key: "spec.source.url", operation: exists }
+          - name: source-from-image-registry
+            selector:
+              matchFields:
+                { key: "spec.source.image", operation: exists }
+
+    - name: source-tester
+      templateRef:
+        kind: ClusterSourceTemplate
+        name: test-source-with-tekton
+      selector:
+        matchLabels:
+          has-tests: "true"
+      sources:
+        - resource: source-provider
+          name: source
+
+    - name: image-provider
+      templateRef:
+        kind: ClusterImageTemplate
+        name: image-from-image-registry
+      selector:
+        matchFields:
+          { key: "spec.image", operation: exists }
+
+    - name: image-builder
+      templateRef:
+        kind: ClusterImageTemplate
+        options:
+          - name: build-image
+            images:
+              - resource: source-tester
+                name: source
+          - name: build-image
+            images:
+              - resource: source-provider
+                name: source
+
+    - name: configure
+      templateRef:
+        kind: ClusterConfigTemplate
+        options:
+          - name: configure
+            sources:
+              - resource: image-builder
+                name: image
+          - name: configure
+            sources:
+              - resource: image-provider
+                name: image
+
+    - name: gitops
+      templateRef:
+        kind: ClusterSourceTemplate
+        options:
+          - name: git-pusher
+            selector:
+              matchLabels:
+                target: gitops
+            configs:
+              - resource: configure
+                name: config
+          - name: registry-pusher
+            selector:
+              matchLabels:
+                target: repostiory
+            configs:
+              - resource: configure
+                name: config
+```
+
+If label `has-tests` is "true", then `image-builder` chooses the first option, from source-tester, because source-tester _will_ eventually be stamped out. 
+Otherwise `image-builder` selects the second option, from source-provider.
+
+This lets a resource be skipped if none of its options match.
+
+`configure` shows a similar pathing, based this time on matchFields not matchLabels.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+An option is selected for one of these reasons:
+
+> It's the only option and all other criteria are met.
+
+OR
+
+> It matches on labels and fields (if no selector is specified, that is a match)
+
+AND
+
+> It's inputs are fulfilled using these same rules.
+
+Options are a priority list to disambiguate inputs that match in more than one case. It should still be possible
```

Comment:
This seems to revert a recent design decision.

In RFC 9 we specify that only 1 option can match. For example, given this workload and supply-chain, an error would be thrown because the conditions for both options are satisfied:

```yaml
kind: Workload
metadata:
  labels:
    is-green: "true"
    is-red: "true"

---
kind: ClusterSupplyChain
spec:
  resources:
    - name: some-step
      templateRef:
        kind: ClusterSourceTemplate
        options:
          - name: option-1
            selector:
              matchFields:
                { key: "metadata.labels.is-green", operation: in, values: ["true"] }
          - name: option-2
            selector:
              matchFields:
                { key: "metadata.labels.is-red", operation: in, values: ["true"] }
```

Making options a priority list reverts that behavior. We see that this same workload paired with a similar supply-chain would now succeed:

```yaml
kind: Workload
metadata:
  labels:
    is-green: "true"
    is-red: "true"

---
kind: ClusterSupplyChain
spec:
  resources:
    - name: some-red-step
      templateRef:
        kind: ClusterSourceTemplate
        name: some-red-template
      selector:
        matchFields:
          { key: "metadata.labels.is-red", operation: in, values: ["true"] }

    - name: some-green-step
      templateRef:
        kind: ClusterSourceTemplate
        name: some-green-template
      selector:
        matchFields:
          { key: "metadata.labels.is-green", operation: in, values: ["true"] }

    - name: some-final-step
      templateRef:
        kind: ClusterSourceTemplate
        options:
          - name: option-1
            sources:
              - resource: some-red-step
                name: source
          - name: option-2
            sources:
              - resource: some-green-step
                name: source
```

The `match one and only one option` rule makes the supply-chain a little more verbose. For example, the image-builder step would become

```yaml
- name: image-builder
  templateRef:
    kind: ClusterImageTemplate
    options:
      - name: build-image
        images:
          - resource: source-tester
            name: source
      - name: build-image
        selector:
          matchFields:
            { key: "metadata.labels.has-test", operation: NotIn, values: ["true"] }  # <--- new selector necessary
        images:
          - resource: source-provider
            name: source
```

Reply:
Image builder being a source tester is obnoxious to me, that's why I prefer to support optional resources.

Reply:
See my comment below about the object encapsulation principle I've used throughout this spec. A resource represents a rigid purpose. `<<interface>> IBuildImages`

Options decorate concrete implementations "<<template>> ImageBuilder" with a selection criteria. 

One purpose is to choose different concrete implementations (eg, `<<template>> GitPusher` and `<<template>> ImagePusher`

The other purpose is to choose the inputs (eg as show in build-image in the example)

These two different purposes for options can be teased apart, but it looks kinda silly because it introduces the `or` syntax that I feel is superfluous (it is essentially the option to add an "or" operator as discussed elsewhere, but it get's very messy when matching Inputs and Fields And Labels.. Trying to absorb that much algebra into a yaml spec is unwieldy) 

So for that reason, I left options selecting either *or both* the concrete class (template) and configuring the inputs.